felix/ipsets/ipsets.go

```diff
@@ -846,74 +803,78 @@ func (s *IPSets) nextFreeTempIPSetName() string {
+		if numDeletions >= MaxIPSetDeletionsPerIteration {
+			// Deleting IP sets is slow (40ms) and serialised in the kernel.  Avoid holding up the main loop
+			// for too long.  We'll leave the remaining sets pending deletion and mop them up next time.
+			log.Debugf("Deleted batch of %d IP sets, rate limiting further IP set deletions.", MaxIPSetDeletionsPerIteration)
+			// Leave the item in the set, so we'll do another batch of deletions next time around the loop.
+			return deltatracker.IterActionNoOpStopIteration
+		}
+		meta, _ := s.setNameToProgrammedMetadata.Dataplane().Get(setName)
+		if meta.DeleteFailed {
+			// We previously failed to delete this IP set, skip it until
+			// the next resync.
+			return deltatracker.IterActionNoOp
+		}
 		logCxt := s.logCxt.WithField("setName", setName)
-		if s.existingIPSetNames.Contains(setName) {
-			logCxt.Info("Deleting IP set.")
-			if err := s.deleteIPSet(setName); err != nil {
-				// Note: we used to set the resyncRequired flag on this path but that can lead to excessive retries if
-				// the problem isn't something that we can fix (for example an external app has made a reference to
-				// our IP set).  Instead, wait for the next timed resync.
-				logCxt.WithError(err).Warning("Failed to delete IP set. Will retry on next resync.")
-			}
+		logCxt.Info("Deleting IP set.")
+		if err := s.deleteIPSet(setName); err != nil {
+			// Note: we used to set the resyncRequired flag on this path but that can lead to excessive retries if
+			// the problem isn't something that we can fix (for example an external app has made a reference to
+			// our IP set).  Instead, wait for the next timed resync.
+			logCxt.WithError(err).Warning("Failed to delete IP set. Will retry on next resync.")
+			meta.DeleteFailed = true
+			s.setNameToProgrammedMetadata.Dataplane().Set(setName, meta)
+			return deltatracker.IterActionNoOp
 		}
-		// Always remove the item so we don't retry until the next timed resync.
-		return set.RemoveItem
+		numDeletions++
+		if _, ok := s.setNameToAllMetadata[setName]; !ok {
+			// IP set is not just filtered out, clean up the members cache.
+			delete(s.mainSetNameToMembers, setName)
+		} else {
+			// We're still tracking this IP set in case it needs to be recreated.
+			// Record that the dataplane is now empty.
+			s.mainSetNameToMembers[setName].Dataplane().DeleteAll()
+		}
+		return deltatracker.IterActionUpdateDataplane
 	})
-
-	// ApplyDeletions() marks the end of the two-phase "apply".  Piggy back on that to
+	// ApplyDeletions() marks the end of the two-phase "apply". Piggyback on that to
 	// update the gauge that records how many IP sets we own.
-	s.gaugeNumIpsets.Set(float64(len(s.ipSetIDToIPSet)))
+	numDeletionsPending := s.setNameToProgrammedMetadata.Dataplane().Len()
+	s.gaugeNumIpsets.Set(float64(numDeletionsPending))
+	return numDeletionsPending > 0 && numDeletions > 0
```

Comment:
` numDeletions > 0` is there to avoid immediate reschedule if there were errors only?

Reply:
Yes, will add comment.

Reply:
what is there was an error but only after some work was completed. Does it make sense to reschedule immediately?